src/compiler/scala/tools/nsc/typechecker/Infer.scala

```diff
@@ -712,6 +712,17 @@ trait Infer extends Checkable {
         argtpes
     }
 
+    // This is primarily a duplicte of enhanceBounds in typedAppliedTypeTree
+    // modified to use updateInfo rather than setInfo to avoid wiping out 
+    // type history.
+    def enhanceBounds(okparams: List[Symbol], okargs: List[Type], undets: List[Symbol]): Unit =
+      undets.foreach { undet =>
+        val bounds = undet.info.bounds
+        val substBounds = bounds.subst(okparams, okargs)
+        if(bounds ne substBounds)
+          undet.updateInfo(substBounds)
```

Comment:
The problem is likely that these sharpened bounds are only intended to be seen for a single candidate in `typedImplicit0`, but end up leaking into later candidates.

Reply:
Yep, seems to be the case:

```
|    |    |    |    [search #7] start `[A, S <: BaseStream[_, S], CC](a: A)(implicit ss: SS[A,S])S`, searching for adaptation to pt=Stream[Int] => IntStream (silent: method x in Test) implicits disabled
|    |    |    |    [search #7] considering f
|    |    |    |    |-- f BYVALmode-EXPRmode-FUNmode-POLYmode (silent: method x in Test) implicits disabled
|    |    |    |    |    [adapt] [A, S <: BaseStream[_, S], CC](a: A)(implicit ss: SS[A,S])S adapted to [A, S <: BaseStream[_, S], CC](a: A)(implicit ss: SS[A,S])S
|    |    |    |    |    \-> (a: A)(implicit ss: SS[A,S])S
|    |    |    |    solving for (A: ?A, S: ?S, CC: ?CC)
enhanceBounds(type S#21256) updated  <: BaseStream#6553[_, S#21256] to  <: BaseStream#6553[_, S#21256]
|    |    |    |    [search #8] start `[A, S <: BaseStream[_, S], CC](a: A)(implicit ss: SS[A,S])S` inferring type S, searching for adaptation to pt=SS[Stream[Int],S] (silent: method x in Test) implicits disabled
|    |    |    |    [search #8] considering anyStreamShape
|    |    |    |    solving for (T: ?T)
|    |    |    |    [adapt] anyStreamShape adapted to [T]=> SS[T,Stream[T]] based on pt SS[Stream[Int],S]
|    |    |    |    [search #8] solve tvars=?S, tvars.constr= >: Stream[Stream[Int]] <: Stream[Stream[Int]]
|    |    |    |    solving for (S: ?S)
enhanceBounds(type S#21256) updated  <: BaseStream#6553[_, S#21256] to  <: BaseStream#6553[_, Stream#6580[Stream#6580[Int#1192]]]
|    |    |    |    [search #8] success inferred value of type SS[Stream[Int],=?Stream[Stream[Int]]] is SearchResult(SSImplicits.anyStreamShape[Stream[Int]], TreeTypeSubstituter(List(type S),List(Stream[Stream[Int]])))
|    |    |    |    [adapt] f adapted to [A, S <: BaseStream[_, S], CC](a: A)(implicit ss: SS[A,S])S based on pt Stream[Int] => IntStream
|    |    |    |    [search #9] start `[A, S <: BaseStream[_, S], CC](a: A)(implicit ss: SS[A,S])S`, searching for adaptation to pt=(=> Stream[Int]) => IntStream (silent: method x in Test) implicits disabled
sandbox/test.scala:23: error: type mismatch;
 found   : Stream[Int]
 required: IntStream
  def x = f(0): IntStream
           ^
```

```diff
⚡ git diff
diff --git a/src/compiler/scala/tools/nsc/typechecker/Infer.scala b/src/compiler/scala/tools/nsc/typechecker/Infer.scala
index 481307467c..4c1994ef61 100644
--- a/src/compiler/scala/tools/nsc/typechecker/Infer.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/Infer.scala
@@ -723,6 +723,9 @@ trait Infer extends Checkable {
         val substBounds = bounds.subst(okparams, okargs)
         if(bounds ne substBounds)
           undet.updateInfo(substBounds)
+        settings.uniqid.value = true
+        println(s"enhanceBounds($undet) updated $bounds to $substBounds")
+        settings.uniqid.value = false
       }
```

Reply:
I'll take a look.

Reply:
Hmm, sounds like we need to clone the undetermined params for each run of the search? Hopefully not too expensive

Reply:
Yup, that was my thought too. I'm surprised I wasn't doing that already.

Reply:
Do you think `undoLog` could be used for this?